src/query/test/test.go

```diff
@@ -0,0 +1,563 @@
+	expected map[uint64]entry
+	m3query  *m3queryClient
+}
+
+type entry struct {
+	pos  int
+	vals []parser.SequenceValue
+}
+
+func (e entry) String() string {
+	return fmt.Sprintf("%d: %s", e.pos, e.vals)
+}
+
+func newEvalCmd(expr string, start time.Time, line int) *evalCmd {
+	return &evalCmd{
+		expr:  expr,
+		start: start,
+		line:  line,
+
+		metrics:  map[uint64]labels.Labels{},
+		expected: map[uint64]entry{},
+		m3query:  newM3QueryClient("localhost", 7201),
+	}
+}
+
+func (ev *evalCmd) String() string {
+	return "eval"
+}
+
+// expect adds a new metric with a sequence of values to the set of expected
+// results for the query.
+func (ev *evalCmd) expect(pos int, m labels.Labels, vals ...parser.SequenceValue) {
+	if m == nil {
+		ev.expected[0] = entry{pos: pos, vals: vals}
+		return
+	}
+	h := m.Hash()
+	ev.metrics[h] = m
+	ev.expected[h] = entry{pos: pos, vals: vals}
+}
+
+const sep = '\xff'
+
+// Hash returns a hash value for the label set.
+func hash(ls model.Metric) uint64 {
+	lbs := make(labels.Labels, 0, len(ls))
+	for k, v := range ls {
+		lbs = append(lbs, labels.Label{
+			Name:  string(k),
+			Value: string(v),
+		})
+	}
+
+	sort.Slice(lbs[:], func(i, j int) bool {
+		return lbs[i].Name < lbs[j].Name
+	})
+
+	return lbs.Hash()
+}
+
+type QueryResponse struct {
+	Status string
+	Data   struct {
+		ResultType string
+		Result     model.Samples
+	}
+}
+
+// compareResult compares the result value with the defined expectation.
+func (ev *evalCmd) compareResult(j []byte) error {
+	var result QueryResponse
+	err := json.Unmarshal(j, &result)
+	if err != nil {
+		return err
+	}
+
+	if result.Status != "success" {
+		return fmt.Errorf("unsuccess status received: %s", result.Status)
+	}
+
+	switch result.Data.ResultType {
+	case "matrix":
+		return errors.New("received range result on instant evaluation")
+
+	case "vector":
+		seen := map[uint64]bool{}
+		for pos, v := range result.Data.Result {
+			fp := hash(v.Metric)
+			if _, ok := ev.metrics[fp]; !ok {
+				return errors.Errorf("unexpected metric %s in result", v.Metric)
+			}
+
+			exp := ev.expected[fp]
+			if ev.ordered && exp.pos != pos+1 {
+				return errors.Errorf("expected metric %s with %v at position %d but was at %d", v.Metric, exp.vals, exp.pos, pos+1)
+			}
+			if !almostEqual(exp.vals[0].Value, float64(v.Value)) {
+				return errors.Errorf("expected %v for %s but got %v", exp.vals[0].Value, v.Metric, v.Value)
+			}
+			seen[fp] = true
+		}
+
+		for fp, expVals := range ev.expected {
+			if !seen[fp] {
+				fmt.Println("vector result", len(result.Data.Result), ev.expr)
+				for _, ss := range result.Data.Result {
+					fmt.Println("    ", ss.Metric, ss.Value)
+				}
+				return errors.Errorf("expected metric %s with %v not found", ev.metrics[fp], expVals)
+			}
+		}
+
+	//case "scalar":
+	// if !almostEqual(ev.expected[0].vals[0].Value, val.V) {
+	// 	return errors.Errorf("expected Scalar %v but got %v", val.V, ev.expected[0].vals[0].Value)
+	// }
```

Comment:
Probably can be uncommented, please check.

Reply:
Agreed to leave it commented as a reference for future use when scalars will be supported.